In [9]:
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd

In [10]:
pip install openpyxl

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [11]:
#PewData Read-in
df = pd.read_csv("/Users/priyachaudhuri/Desktop/AI-Human-PewData.csv")
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10260 entries, 0 to 10259
Columns: 215 entries, Sl to WEIGHT_W99
dtypes: float64(2), int64(1), object(212)
memory usage: 16.8+ MB


## There are 215 columns in total.

### The main focus of our proposal is on Chip Implant and Gene Editing, therefore, we are going to clean the dataset by dropping the columns that are not necessary to our main focus. 

In [12]:
# Drop the Interview start & end columns and Device type as they are less relevant to our group's focus 
df.drop(df.iloc[:, 2:5], inplace = True, axis = 1)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10260 entries, 0 to 10259
Columns: 212 entries, Sl to WEIGHT_W99
dtypes: float64(2), int64(1), object(209)
memory usage: 16.6+ MB


In [13]:
# Drop tablet type and all other AI technology groups question codes (e.g. facial recognition technology, 
# driverless passenger vehicle)
# keep questions only for Gene Editing and Chip Implant technologies (our main focus)

df.drop(df.columns[[4] + list(range(35, 110)) + list(range(158, 179))], inplace = True, axis = 1)
df.info()

# Now the dataset contains columns that are necessary to our proposal focus 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10260 entries, 0 to 10259
Columns: 115 entries, Sl to WEIGHT_W99
dtypes: float64(2), int64(1), object(112)
memory usage: 9.0+ MB


In [14]:
# Check for missing values -- Notice there are missing values under question codes BCHIP  
bchip_columns = df.columns[df.columns.str.contains('BCHIP')]
null_counts = df[bchip_columns].isnull().sum()
null_counts

BCHIP1_W99       5153
BCHIP2_W99       5153
BCHIP3_W99       5153
BCHIP4_a_W99     5153
BCHIP4_b_W99     5153
BCHIP4_c_W99     5153
BCHIP4_d_W99     5153
BCHIP5_W99       5153
BCHIP6_W99       5153
BCHIP7_W99       5153
BCHIP8_W99       5153
BCHIP9_a_W99     5153
BCHIP9_b_W99     5153
BCHIP9_c_W99     5153
BCHIP10_a_W99    5153
BCHIP10_b_W99    5153
BCHIP10_c_W99    5153
BCHIP10_d_W99    5153
BCHIP11_W99      5153
BCHIP12_W99      5153
BCHIP13_a_W99    5153
BCHIP13_b_W99    5153
BCHIP14_a_W99    5153
BCHIP14_b_W99    5153
BCHIP14_c_W99    5153
BCHIP14_d_W99    5153
dtype: int64

In [15]:
# Check for missing values -- Notice there are missing values under question codes GENEV  
gene_columns = df.columns[df.columns.str.contains('GENEV')]
df[gene_columns].isnull().sum()

GENEV1_W99       5153
GENEV2_W99       5153
GENEV3_W99       5153
GENEV4_a_W99     5153
GENEV4_b_W99     5153
GENEV4_c_W99     5153
GENEV4_d_W99     5153
GENEV5_W99       5153
GENEV6_W99       5153
GENEV7_W99       5153
GENEV8_W99       5153
GENEV9_a_W99     5153
GENEV9_b_W99     5153
GENEV9_c_W99     5153
GENEV9_d_W99     5153
GENEV10_W99      5153
GENEV11_W99      5153
GENEV12_a_W99    5153
GENEV12_b_W99    5153
GENEV12_c_W99    5153
GENEV13_a_W99    5153
GENEV13_b_W99    5153
dtype: int64

In [16]:
# There are in total 5153 missing values under each column of BCHIP - 
# indicating 5153 participants did not answer any question of Chip Implant in this questionnaire, 
# we should consider dropping these rows since they did not provide any insight of Chip Implant
df_clean_1 = df.dropna(subset = bchip_columns)
df_clean_1[bchip_columns].isnull().sum()

BCHIP1_W99       0
BCHIP2_W99       0
BCHIP3_W99       0
BCHIP4_a_W99     0
BCHIP4_b_W99     0
BCHIP4_c_W99     0
BCHIP4_d_W99     0
BCHIP5_W99       0
BCHIP6_W99       0
BCHIP7_W99       0
BCHIP8_W99       0
BCHIP9_a_W99     0
BCHIP9_b_W99     0
BCHIP9_c_W99     0
BCHIP10_a_W99    0
BCHIP10_b_W99    0
BCHIP10_c_W99    0
BCHIP10_d_W99    0
BCHIP11_W99      0
BCHIP12_W99      0
BCHIP13_a_W99    0
BCHIP13_b_W99    0
BCHIP14_a_W99    0
BCHIP14_b_W99    0
BCHIP14_c_W99    0
BCHIP14_d_W99    0
dtype: int64

In [17]:
df_clean_1[gene_columns].isnull().sum()    #dataset that have Chip Implant questions cleaned 

# Use this cleaned dataset to check if there's still any missing values for Gene Editing columns 
# Notice there are none, indicating those 5153 participants who did not answer questions for Chip implant 
# also did not answer questions for Gene Editing 

# This is due to the reason all Form 1 respondants are only supposed to answer question regarding 
# social media companies, facial recognition and driveless passenger vehicle 

# Only Form 2 respondants got questions regarding Gene Editing and Chip Implants 

GENEV1_W99       0
GENEV2_W99       0
GENEV3_W99       0
GENEV4_a_W99     0
GENEV4_b_W99     0
GENEV4_c_W99     0
GENEV4_d_W99     0
GENEV5_W99       0
GENEV6_W99       0
GENEV7_W99       0
GENEV8_W99       0
GENEV9_a_W99     0
GENEV9_b_W99     0
GENEV9_c_W99     0
GENEV9_d_W99     0
GENEV10_W99      0
GENEV11_W99      0
GENEV12_a_W99    0
GENEV12_b_W99    0
GENEV12_c_W99    0
GENEV13_a_W99    0
GENEV13_b_W99    0
dtype: int64

In [18]:
df_clean_1.describe(include = 'all')

# The dataset now has 5107 rows to analyze

Sl          QKEY LANG_W99 FORM_W99 TECH1_W99  \
count    5107.000000  5.107000e+03     5107     5107         0   
unique           NaN           NaN        2        1         0   
top              NaN           NaN  English   Form 2       NaN   
freq             NaN           NaN     4874     5107       NaN   
mean     5129.626395  1.730272e+11      NaN      NaN       NaN   
std      2970.325559  7.061263e+10      NaN      NaN       NaN   
min         0.000000  1.002600e+05      NaN      NaN       NaN   
25%      2570.500000  2.017020e+11      NaN      NaN       NaN   
50%      5150.000000  2.018010e+11      NaN      NaN       NaN   
75%      7742.500000  2.019010e+11      NaN      NaN       NaN   
max     10256.000000  2.021010e+11      NaN      NaN       NaN   

                SC1_W99                     CNCEXC_W99  \
count              5107                           5107   
unique                4                              4   
top     Mostly positive  Equally concerned and excited   
freq               3651                           2330   
mean                NaN                            NaN   
std                 NaN                            NaN   
min                 NaN                            NaN   
25%                 NaN                            NaN   
50%                 NaN                            NaN   
75%                 NaN                            NaN   
max                 NaN                            NaN   

                  EXCITEOE_W99_OE1            EXCITEOE_W99_OE2  \
count                          964                         161   
unique                          13                          11   
top     Makes life, society better  Saves time, more efficient   
freq                           286                          36   
mean                           NaN                         NaN   
std                            NaN                         NaN   
min                            NaN                         NaN   
25%                            NaN                         NaN   
50%                            NaN                         NaN   
75%                            NaN                         NaN   
max                            NaN                         NaN   

             EXCITEOE_W99_OE3  ...       F_PARTYLN_FINAL F_PARTYSUM_FINAL  \
count                      29  ...                  1864             5107   
unique                      8  ...                     3                3   
top     Helps with work/labor  ...  The Democratic Party     Dem/Lean Dem   
freq                        9  ...                   872             2553   
mean                      NaN  ...                   NaN              NaN   
std                       NaN  ...                   NaN              NaN   
min                       NaN  ...                   NaN              NaN   
25%                       NaN  ...                   NaN              NaN   
50%                       NaN  ...                   NaN              NaN   
75%                       NaN  ...                   NaN              NaN   
max                       NaN  ...                   NaN              NaN   

         F_PARTYSUMIDEO_FINAL        F_INC_SDT1  \
count                    5107              5107   
unique                      5                10   
top     Conservative Rep/Lean  $100,000 or more   
freq                     1711              1447   
mean                      NaN               NaN   
std                       NaN               NaN   
min                       NaN               NaN   
25%                       NaN               NaN   
50%                       NaN               NaN   
75%                       NaN               NaN   
max                       NaN               NaN   

                                                    F_REG    F_IDEO  \
count                                                4893      5107   
unique                                                 

## Separate the entire dataset into two datasets.
### One focusing on Chip Implant and one focusing on Gene Editing 

In [19]:
# Identifying the general columns -- demographic attributes columns (the ones that do not contain
# BCHIP or GENEV) 
general_columns = df_clean_1.columns[~df_clean_1.columns.str.contains('BCHIP|GENEV')]

In [20]:
df_bchip = df_clean_1[general_columns.union(bchip_columns)]
df_bchip.info()

# in this dataframe -- contains only Chip Implant questions and generic information (demographic attributes)
# by doing so, easier for the group to analyze these two technologies separately 

<class 'pandas.core.frame.DataFrame'>
Index: 5107 entries, 0 to 10256
Data columns (total 93 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   ALGFAIR_W99            5107 non-null   object 
 1   BCHIP10_a_W99          5107 non-null   object 
 2   BCHIP10_b_W99          5107 non-null   object 
 3   BCHIP10_c_W99          5107 non-null   object 
 4   BCHIP10_d_W99          5107 non-null   object 
 5   BCHIP11_W99            5107 non-null   object 
 6   BCHIP12_W99            5107 non-null   object 
 7   BCHIP13_a_W99          5107 non-null   object 
 8   BCHIP13_b_W99          5107 non-null   object 
 9   BCHIP14_a_W99          5107 non-null   object 
 10  BCHIP14_b_W99          5107 non-null   object 
 11  BCHIP14_c_W99          5107 non-null   object 
 12  BCHIP14_d_W99          5107 non-null   object 
 13  BCHIP1_W99             5107 non-null   object 
 14  BCHIP2_W99             5107 non-null   object 
 15  BCHIP3_W

In [21]:
#Export the Chip dataset into Excel
df_bchip.to_excel("output_bchip.xlsx",
             sheet_name='BCHIP') 

In [22]:
df_genev = df_clean_1[general_columns.union(gene_columns)]
df_genev.info()

# in this dataframe -- contains only Gene Editinng questions and generic information (demographic attributes)
# by doing so, easier for the group to analyze these two technologies separately 

<class 'pandas.core.frame.DataFrame'>
Index: 5107 entries, 0 to 10256
Data columns (total 89 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   ALGFAIR_W99            5107 non-null   object 
 1   CNCEXC_W99             5107 non-null   object 
 2   CONCERNOE_REFUSED_W99  5107 non-null   object 
 3   CONCERNOE_W99_OE1      1792 non-null   object 
 4   CONCERNOE_W99_OE2      325 non-null    object 
 5   CONCERNOE_W99_OE3      51 non-null     object 
 6   DISCRIM1_a_W99         5107 non-null   object 
 7   DISCRIM1_b_W99         5107 non-null   object 
 8   DISCRIM1_c_W99         5107 non-null   object 
 9   DISCRIM1_d_W99         5107 non-null   object 
 10  DISCRIM1_e_W99         5107 non-null   object 
 11  DISCRIM1_f_W99         5107 non-null   object 
 12  EXCITEOE_REFUSED_W99   5107 non-null   object 
 13  EXCITEOE_W99_OE1       964 non-null    object 
 14  EXCITEOE_W99_OE2       161 non-null    object 
 15  EXCITEOE

In [23]:
#Export the Gene Editing dataset into Excel
df_genev.to_excel("output_genev.xlsx",
             sheet_name='GENEV') 

## Data Summaries for Validation

In [24]:
df_bchip.groupby('F_AGECAT')['QKEY'].count().sort_values()

F_AGECAT
Refused      18
18-29       438
50-64      1444
65+        1521
30-49      1686
Name: QKEY, dtype: int64

In [25]:
df_bchip.groupby('F_GENDER')['QKEY'].count().sort_values()

F_GENDER
Refused                11
In some other way      29
A man                2279
A woman              2788
Name: QKEY, dtype: int64

In [26]:
df_bchip.groupby('F_RACECMB')['QKEY'].count().sort_values()


F_RACECMB
Refused                        99
Or some other race            177
Asian or Asian-American       190
Mixed Race                    190
Black or African-American     427
White                        4024
Name: QKEY, dtype: int64

In [27]:
df_bchip.groupby('F_EDUCCAT2')['QKEY'].count().sort_values()

F_EDUCCAT2
Refused                              16
Less than high school               166
Associate's degree                  553
High school graduate                703
Some college, no degree            1066
Postgraduate                       1222
College graduate/some post grad    1381
Name: QKEY, dtype: int64

In [28]:
df_bchip.groupby('F_RELIGCAT1')['QKEY'].count().sort_values()

F_RELIGCAT1
Refused           31
Other            461
Catholic        1116
Unaffiliated    1383
Protestant      2116
Name: QKEY, dtype: int64

In [29]:
df_bchip.groupby('F_INC_TIER2')['QKEY'].count().sort_values()

F_INC_TIER2
Refused           264
Lower income     1094
Upper income     1248
Middle income    2501
Name: QKEY, dtype: int64